### ISSUE TITLE
Failure during update metadata

### ISSUE THREAD

**Issue by rexendevar** [comixed/comixed#2203]
**Describe The Bug**
Running a fairly large job on a bunch of files. Suddenly it stops saving the metadata to the files. No idea why. It's 1 am and I don't really have time to troubleshoot currently but it seems to be doing it on multiple files??? The message isn't telling me what book it's trying to apply to so I don't know.

**To Reproduce The Bug**
no idea sorry

**ComiXed Runtime Environment**

 - [ ] Windows
 - [x] Linux (any distro)
 - [ ] MacOS
 - [ ] Docker (if so, please select the host OS)

**Browser (If A Web Issue)**

 - [ ] Chrome
 - [x] Firefox
 - [ ] Safari
 - [ ] Other (if so, please provide a name)

**ComiXed Build Detail:**
```
The code was built from a branch named 6a388a7644f82f27f3ebe71241e7aad626e70a5d
The code was built on 2024-11-10T06:04:49.000+00:00.
Code was built on a machine named fv-az700-429.
The code was marked with a Build version of 2.2.2-1.
The latest commit hash on the source branch is 6a388a7644f82f27f3ebe71241e7aad626e70a5d.
This commit was created on 2024-11-10T06:00:43.000+00:00.
Commit Message
Changed the release version to v2.2.2-1 [#2157]
Commit was created by Darryl L. Pierce <mcpierce@gmail.com>
The branch does not contain uncommitted changes.
This branch is tracking the remote branch https://github.com/comixed/comixed
The database URL being used is jdbc:h2:file:~/.comixed/comixed
Using Java 21.0.5+11-LTS (Azul Systems, Inc.) running on Linux v6.9.3-76060903-generic (amd64)
```

**Additional Context**

**Error Message**
```
  java.lang.NullPointerException: Cannot invoke "org.comixedproject.model.comicpages.ComicPage.getFilename()" because "page" is null
	at org.comixedproject.adaptors.comicbooks.ComicBookAdaptor.save(ComicBookAdaptor.java:178)
	at org.comixedproject.batch.comicbooks.processors.UpdateMetadataProcessor.process(UpdateMetadataProcessor.java:71)
	at java.base/jdk.internal.reflect.DirectMethodHandleAccessor.invoke(DirectMethodHandleAccessor.java:103)
	at java.base/java.lang.reflect.Method.invoke(Method.java:580)
	at org.springframework.aop.support.AopUtils.invokeJoinpointUsingReflection(AopUtils.java:351)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.invokeJoinpoint(ReflectiveMethodInvocation.java:196)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:163)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:765)
	at org.springframework.aop.support.DelegatingIntroductionInterceptor.doProceed(DelegatingIntroductionInterceptor.java:137)
	at org.springframework.aop.support.DelegatingIntroductionInterceptor.invoke(DelegatingIntroductionInterceptor.java:124)
	at org.springframework.aop.framework.ReflectiveMethodInvocation.proceed(ReflectiveMethodInvocation.java:184)
	at org.springframework.aop.framework.CglibAopProxy$CglibMethodInvocation.proceed(CglibAopProxy.java:765)
	at org.springframework.aop.framework.CglibAopProxy$DynamicAdvisedInterceptor.intercept(CglibAopProxy.java:717)
	at org.comixedproject.batch.comicbooks.processors.UpdateMetadataProcessor$$SpringCGLIB$$0.process(<generated>)
	at org.springframework.batch.core.step.item.SimpleChunkProcessor.doProcess(SimpleChunkProcessor.java:146)
	at org.springframework.batch.core.step.item.SimpleChunkProcessor.transform(SimpleChunkProcessor.java:322)
	at org.springframework.batch.core.step.item.SimpleChunkProcessor.process(SimpleChunkProcessor.java:220)
	at org.springframework.batch.core.step.item.ChunkOrientedTasklet.execute(ChunkOrientedTasklet.java:75)
	at org.springframework.batch.core.step.tasklet.TaskletStep$ChunkTransactionCallback.doInTransaction(TaskletStep.java:388)
	at org.springframework.batch.core.step.tasklet.TaskletStep$ChunkTransactionCallback.doInTransaction(TaskletStep.java:312)
	at org.springframework.transaction.support.TransactionTemplate.execute(TransactionTemplate.java:140)
	at org.springframework.batch.core.step.tasklet.TaskletStep$2.doInChunkContext(TaskletStep.java:255)
	at or
  ```

**Comment by cjil** [comment]
Interestingly, I had the same error message the other day, though I believe in my case it was self-inflicted as I was manually editing the database to change the filetype to cbz, after running a bash script to convert all cbr to cbz after discovering that the comicinfo.xml file is not updated in cbr files, and I was unable to add all 9000+ cbr files to a single batch process for conversion.

In the end I nuked my database and started again. Hopefully you have a better time of it than me.

**Comment by mcpierce** [comment]
@rexendevar Prior to the stacktrace, there should be some log entries that names the file being processed. Is there anything about that file that may be an issue? If you try to update that file directly does it fail consistently? Could you possibly send me the file so I can look at what might be an issue with it?

**Comment by rexendevar** [comment]
![image](https://github.com/user-attachments/assets/ea475d3d-0f9e-4ce7-b6cf-81435f517f43)
it's this for ages and ages. I can give you the file but I wanna do it privately because it's copyrighted - how can i send it to you for testing?

**Comment by rexendevar** [comment]
If it's relevant, the "update metadata process" tab has been blank the entire time

**Comment by rexendevar** [comment]
I have removed the file from the queue and it successfully wrote a few files, then stopped again

**Comment by rexendevar** [comment]
It has now written metadata to one comic multiple times

**Comment by rexendevar** [comment]
I uzipped and rezipped the file in question and it didn't fix anything.

**Comment by mcpierce** [comment]
> If it's relevant, the "update metadata process" tab has been blank the entire time

I think the root problem is likely just that the batch process got into a bad state and wasn't able to recover. So I don't think the file itself was the issue here. 

Try this: go to the Batch Processes page and make sure that all completed processes are cleared out, and then select and kill any that remain. Then try updating a few comics' metadata and see what happens.

I'm verifying this behavior locally as well: I have a test server with about 30k comics, rescraped about 300 of them, and then ran the update metadata batch process. It's relatively slow, but it completed for me without incident.

That said, if we can find a way to consistent reproduce the problem, we'll definitely find a way to fix it.

**Comment by rexendevar** [comment]
How do I select and kill processes in the batch processes page? All I can do is hover over them.

**Comment by mcpierce** [comment]
> How do I select and kill processes in the batch processes page? All I can do is hover over them.

I think you're looking at the Processing Comics page and not the Batch Processes page:
![image](https://github.com/user-attachments/assets/e7946aca-3968-43c5-a4fe-c6dd30baee82)

Select the ones to kill:
![image](https://github.com/user-attachments/assets/9e3c5bad-e304-4454-871c-e8c7e9a69d48)

And click the Kill action button:
![image](https://github.com/user-attachments/assets/b0c0c56d-ca76-4f8d-9242-a1cf3282dd6d)

**Comment by rexendevar** [comment]
no I'm dumb, I thought you were talking about the Processing Comics page. That said, there is a job in the Processing Comics page that doesn't exist in the Batch Processes page. 
![image](https://github.com/user-attachments/assets/7527bbec-1165-4aae-91aa-2b6171446448)
![image](https://github.com/user-attachments/assets/d6910c50-776c-47c7-9db4-d7bb8d0a908e)
because I can't find the continuous update metadata job in the batch processes page, I'm not able to kill it and restart it.

**Comment by mcpierce** [comment]
If the process isn't a ghost, it'll be on the batch processes page. Try sorting the processes by *Started* descending and see if it's closer to the top.

**Comment by rexendevar** [comment]
The only thing I'm seeing is the loading page hashes job, and these update metadata jobs that failed. I cannot see the one that is still trying to run.

**Comment by mcpierce** [comment]
> The only thing I'm seeing is the loading page hashes job, and these update metadata jobs that failed. I cannot see the one that is still trying to run.

The Batch Processes page is showing data directly from the batch processes services, which is reading them from the database as well as from memory. That's why it'll show processes that have already completed.

Have you tried reloading the page to see if that clears the Processing Comics page? That one only shows output when batch processes publish updates as they complete steps. If you reload that page and it doesn't reshow that process then it was just a ghost record that was in the browser but dead and gone from the backend. If it shows status then it literally *has* to be running on the back end and should show up in the Batch Processes page.

**Comment by rexendevar** [comment]
Reloading seems to have fixed it. I'm going to try updating the metadata again.

**Comment by rexendevar** [comment]
Spoke too soon, the update comic book metadata process has reappeared.

**Comment by rexendevar** [comment]
The in-progress update metadata job does not show up on the batch processes page. It only shows up when it fails.

**Comment by mcpierce** [comment]
> The in-progress update metadata job does not show up on the batch processes page. It only shows up when it fails.

All batch processes publish updates that show up on that page. Not all batch processes start immediately after you click on them, though.  Currently the initiators scan the database periodically and look for conditions to start the batch processes. We've got a feature request to connect the menu options to the initiators so they are instantly started, but haven't had a chance to implement that yet.

(edit)

Correction: the batch processes do have the initiators invoked when the option is selected. I forgot that had been done. But the batch processes don't always start right away, even when invoked. The Spring batch library runs the processes in parallel to the rest of the server's functionality, so it can sometimes be lower priority if something else is going on, like browser activity, which requires higher priority.

**Comment by rexendevar** [comment]
Is there a file I can delete to manually remove all in-progress jobs?

**Comment by mcpierce** [comment]
> Is there a file I can delete to manually remove all in-progress jobs?

No, there's not. All of the batch process data is in the database. You can clear them out by going to the Batch Processes page, selecting all processes, and then clocking the delete selected processes button to clear them out.

**Comment by mcpierce** [comment]
@rexendevar I *think* there might be some odd interplay between the batch and state machine handling. I'm trying something out with the next release that I'm hoping resolves the issue. We can keep this bug open and see if, over the next few releases, the issue doesn't reappear. If it doesn't, we can close it later.

**Comment by rexendevar** [comment]
Beginner question: I'm trying to build the repository and it's failing. I realize I'm supposed to add to the developer mailing list when I'm dealing with a bug in a pre-release instead of a full release. However, I have never dealt with a mailing list before and I do not know how to send anything to it. Please help me

**Comment by mcpierce** [comment]
> Beginner question: I'm trying to build the repository and it's failing. I realize I'm supposed to add to the developer mailing list when I'm dealing with a bug in a pre-release instead of a full release. However, I have never dealt with a mailing list before and I do not know how to send anything to it. Please help me

Go to this URL and subscribe:

https://www.freelists.org/list/comixed-dev

Then you would just send an email to comixed-dev@freelists.org to post to the mailing list. But you have to subscribe to it first since only members can post to the list.
